File: app/src/main/java/com/notes/ayatkhraisat/mvppratctice/di/ApplicationModule.java
package com.notes.ayatkhraisat.mvppratctice.di;

import android.app.Application;
import android.content.Context;
import com.notes.ayatkhraisat.mvppratctice.App;
import dagger.Binds;
import dagger.Module;
import dagger.Provides;

import javax.inject.Singleton;

@Module
public class ApplicationModule {

    private Application application;


    ApplicationModule(Application application) {
        this.application = application;
    }

    @Singleton
    @Provides
    Context bindContext(Application application) {
        return application;
    }
}

File: app/src/main/java/com/ayatkhraisat/mvp_example/di/modules/ApplicationModule.kt
package com.ayatkhraisat.mvp_example.di.modules

import android.content.Context
import com.ayatkhraisat.mvp_example.data.room.AppDatabase
import com.ayatkhraisat.mvp_example.data.room.MovieDao
import com.ayatkhraisat.mvp_example.data.room.SharedPreferencesHelper
import com.ayatkhraisat.mvp_example.di.qualifires.ApplicationContext
import com.ayatkhraisat.mvp_example.di.scopes.ViewScope
import dagger.Module
import dagger.Provides
import javax.inject.Singleton

/**
 * Created by Ayat Khriasat on 14,December,2018 at 7:33 PM
 * Email: ayatzkhraisat@gmail.com
 * Project: MvpExample
 */
@Module
class ApplicationModule(val application: Context) {


    @Singleton
    @ApplicationContext
    @Provides
    fun provideApplicationContext(): Context = application


    @Singleton
    @Provides
    fun provideSharedPreferences(@ApplicationContext application: Context): SharedPreferencesHelper =
        SharedPreferencesHelper(application);

    @Provides
    @Singleton
    fun provideMovieDao(@ApplicationContext application: Context): MovieDao {
        return AppDatabase.getAppDataBase(application)!!.taskDao()

    }

}
